### ISSUE TITLE
Mouse cursor invisible -- MacOS

### ISSUE THREAD

**Issue by knubie** [kovidgoyal/kitty#1218]
### Steps to reproduce:
- Open kitty (maximize the window so it fills up most of the screen)
- Open some other app (like finder)
- Use kitty until the mouse cursor disappears.
- Cmd+tab to the other app such that the mouse cursor is already positioned on the other app window.
   - Note that the mouse cursor is now visible.
- Move the mouse cursor off of the other app window and on top of the kitty window.
   - Note that the mouse cursor will now disappear.
- Move the mouse cursor back onto the other app window.
   - Note that the mouse cursor is still hidden

### Expected behavior:
I would not expect kitty to ever hide the mouse cursor when it does not have focus. I would also not expect it to hide a moving mouse cursor.

It seems that kitty is too aggressively hiding the mouse cursor, even when it doesn't have focus.

**Comment by lithammer** [comment]
Does setting `mouse_hide_wait` to `0` alleviate the issue?

**Comment by Luflosi** [comment]
@lithammer pretty sure, that your suggestion will work around the issue for him.
@knubie Could you try to recompile kitty with https://github.com/Luflosi/kitty/commit/362ab966ce1d9b6261e03257a932de49ebf79044 applied? It might fix your issue without needing to disable mouse cursor hiding.

**Comment by kovidgoyal** [comment]
I'm somewhat surprised that macos would even allow non focused applications to hide the mouse cursor. That seems like a security issue. Non focused applications should not be able to control the mouse cursor from under the focused application, at least without special privileges.

**Comment by knubie** [comment]
It's actually difficult to reproduce this when kitty first loads. Only once the bug happens "naturally" am I able to consistently reproduce it. I'm sure exactly what's triggering it.

**Comment by kovidgoyal** [comment]
I cannot reproduce, and looking at the code, I dont see anything wrong that kitty is doing. There are only two places where kitty asks the OS to hide the mouse cursor. 

1) When cocoa fires the mouseEnter event and the cursor is supposed to be hidden (this is the code path @Luflosi is referencing). There is a corresponding showCursor call in mouseExit.

2) When the mouse hide timeout expires, here kitty calls glfwSetInputMode, which calls _glfwPlatformSetCursorMode which in turn only does anything if _glfwPlatformWindowIsFocused returns true (i.e. if the cocoa isKeyWindow method returns true which means the window is focused.

So kitty never asks the OS to hide the mouse unless the window has focus. I'm guessing the issue is because mouseEntered/mouseExited is not being sent by cocoa when you dont actually move the mouse but instead use cmd-tab to switch applications. Not sure what can be done about that. Perhaps this patch, which I cannot test since I cannot reproduce:

```patch
diff --git a/glfw/cocoa_window.m b/glfw/cocoa_window.m
index b6c63488..7305159c 100644
--- a/glfw/cocoa_window.m
+++ b/glfw/cocoa_window.m
@@ -455,6 +455,7 @@ static const NSRange kEmptyRange = { NSNotFound, 0 };
 
     _glfwInputWindowFocus(window, GLFW_TRUE);
     updateCursorMode(window);
+    if (window->cursorMode == GLFW_CURSOR_HIDDEN) hideCursor(window);
     if (_glfw.ns.disabledCursorWindow != window && cursorInClientArea(window))
     {
         double x = 0, y = 0;
@@ -467,6 +468,7 @@ static const NSRange kEmptyRange = { NSNotFound, 0 };
 {
     if (window->monitor && window->autoIconify)
         _glfwPlatformIconifyWindow(window);
+    showCursor(window);
 
     _glfwInputWindowFocus(window, GLFW_FALSE);
 }
@@ -713,17 +715,11 @@ static GLFWapplicationshouldhandlereopenfun handle_reopen_callback = NULL;
 
 - (void)mouseExited:(NSEvent *)event
 {
-    if (window->cursorMode == GLFW_CURSOR_HIDDEN)
-        showCursor(window);
-
     _glfwInputCursorEnter(window, GLFW_FALSE);
 }
 
 - (void)mouseEntered:(NSEvent *)event
 {
-    if (window->cursorMode == GLFW_CURSOR_HIDDEN)
-        hideCursor(window);
-
     _glfwInputCursorEnter(window, GLFW_TRUE);
 }
```

**Comment by Luflosi** [comment]
I'll try your patch for a week or so.

**Comment by knubie** [comment]
Thanks for the patch @koekeishiya, I will keep using the un-patched version to see if I can figure out steps to reliable reproduce the bug.

**Comment by knubie** [comment]
Haven't noticed this issue in a few days, so I'll close the issue.

**Comment by Luflosi** [comment]
Or maybe we need a check in `if (OPT(mouse_hide_wait) > 0 && !is_mouse_hidden(os_window)) {` in kitty/child-monitor.c if the window has focus.

**Comment by kovidgoyal** [comment]
last_mouse_activity is updated when there is any mouse activity at all, enter/leave, button press, movement.

As I said right at the beginning, no unfocused application should be able to change the mouse cursor visibility, no matter what it does, as that is a security risk. And I cannot reproduce this issue. So simply flailing in the dark doesn't seem like a very effective way to fix this.

**Comment by octplane** [comment]
I also see the same issue, although I did not found any way to reproduce it. It happens randomly but quite often and my recent increased usage of kitty suggests relation between the two. 🤷‍♂️

**Comment by knubie** [comment]
I've also started noticing it again, but it's very difficult to reproduce.

**Comment by amarshall** [comment]
I see this behavior *a lot* (it drives me nuts, hadn’t realized it might be Kitty), and I don’t know how to reproduce it *starting* to happen, either (but once it starts, the original reproduction steps work). I’m going to run Kitty with the above patch for a week and report back.

**Comment by DealPete** [comment]
This issue should not be closed, it happens frequently to me as well.

**Comment by amarshall** [comment]
I used the [above patch](https://github.com/kovidgoyal/kitty/issues/1218#issuecomment-446074463) applied to v0.13.3 all week without exiting Kitty on two different machines both running macOS 10.13.6. Didn’t have the cursor disappearing issue once. At the end, I decided to try the unpatched v0.13.3 for a bit, and within an hour had the cursor disappearance issue. Further, there didn’t appear to be any issues functionally with the normal behavior of the cursor disappearing after the configured time period when Kitty has focus.

As such, it’d be great to see the patch merged and released, as it appears to resolve the issue.
